transport/src/test/java/io/netty/channel/embedded/EmbeddedChannelTest.java

```diff
@@ -380,6 +384,161 @@ public void run() {
+
+      // There was no #flushInbound() call so nobody should have called
+      // #channelReadComplete()
+      assertEquals(flushCount.get(), 0);
+    }
+
+    @Test
+    public void testFlushOutbound() throws InterruptedException {
+        final CountDownLatch latch = new CountDownLatch(1);
+        EmbeddedChannel channel = new EmbeddedChannel(new ChannelOutboundHandlerAdapter() {
+            @Override
+            public void flush(ChannelHandlerContext ctx) throws Exception {
+                latch.countDown();
+            }
+        });
+
+        channel.flushOutbound();
+
+        if (!latch.await(1L, TimeUnit.SECONDS)) {
+            fail("Nobody called #flush() in time.");
+        }
+    }
+
+    @Test
+    public void testWriteOneOutbound() throws InterruptedException {
+        final CountDownLatch latch = new CountDownLatch(1);
+        final AtomicInteger flushCount = new AtomicInteger(0);
+
+        EmbeddedChannel channel = new EmbeddedChannel(new ChannelOutboundHandlerAdapter() {
+            @Override
+            public void write(ChannelHandlerContext ctx, Object msg, ChannelPromise promise) throws Exception {
+                latch.countDown();
+                ctx.write(msg, promise);
+            }
+
+            @Override
+            public void flush(ChannelHandlerContext ctx) throws Exception {
+                flushCount.incrementAndGet();
+            }
+        });
+
+        channel.writeOneOutbound("Hello, Netty!");
+
+        if (!latch.await(1L, TimeUnit.SECONDS)) {
+            fail("Nobody called #write() in time.");
+        }
+
+        channel.close().syncUninterruptibly();
+
+        // There was no #flushOutbound() call so nobody should have called
+        // #flush()
+        assertEquals(flushCount.get(), 0);
+    }
+
+    @Test
+    public void testEnsureOpen() throws InterruptedException {
+        EmbeddedChannel channel = new EmbeddedChannel();
+        channel.close().syncUninterruptibly();
+
+        try {
+            channel.writeOutbound("Hello, Netty!");
+            fail("This should have failed with a ClosedChannelException");
+        } catch (Exception expected) {
+            assertTrue(expected instanceof ClosedChannelException);
+        }
+
+        try {
+            channel.writeInbound("Hello, Netty!");
+            fail("This should have failed with a ClosedChannelException");
+        } catch (Exception expected) {
+            assertTrue(expected instanceof ClosedChannelException);
+        }
+    }
+
+    @Test
+    public void testHandleInboundMessage() throws InterruptedException {
+        final CountDownLatch latch = new CountDownLatch(1);
+
+        EmbeddedChannel channel = new EmbeddedChannel() {
+            @Override
+            protected void handleInboundMessage(Object msg) {
+                latch.countDown();
+            }
+        };
+
+        channel.writeOneInbound("Hello, Netty!");
+
+        if (!latch.await(1L, TimeUnit.SECONDS)) {
```

Comment:
EmbeddedChannel uses something like an ImmediateEventLoop. The `#writeOneInbound()` should immediately call `#handleInboundMessage()`.